### PR TITLE
add support for services

### PR DESCRIPTION
**Pull request by galdor**

### CODE REVIEW

**Comment by clutchski** [comment]
why do we vendor grpc here?

**Comment by galdor** [comment]
Because this repository uses a recent version of grpc (`grpc.SupportPackageIsVersion4`), while some of our other repositories use an older one (`dd-go` requires `grpc.SupportPackageIsVersion3`). `dd-trace-go` does not use any dependency manager, so we have to either vendor as I do in this commit, or manually upgrade/downgrade the grpc repository in `$GOPATH` every time we switch to another repository.

Note that switching from glock to glide for all repositories would solve this problem, since glide installs versionned dependencies in each project `vendor` directory instead of modifying packages in `$GOPATH`.

**Comment by LeoCavaille** [comment]
I guess not vendoring it was an oversight of the initial work on grpc.

**Comment by LeoCavaille** [comment]
Misread the grpc thing, will chat with Nicolas about vendoring it.

File: tracer/tracer.go [hunk 1]
```diff
@@ -23,6 +34,13 @@ type Tracer struct {
 
 	DebugLoggingEnabled bool
 	enabled             bool // defines if the Tracer is enabled or not
+
+	services         map[string]*Service // name -> service
+	servicesModified bool
+	serviceChan      chan *Service
```
Comment (LeoCavaille) [inline tracer/tracer.go]: Here not sure the level of pointer indirection is worth it, any particular reason except tiny perf boost?

File: tracer/tracer.go [hunk 2]
```diff
@@ -39,6 +57,12 @@ func NewTracerTransport(transport Transport) *Tracer {
 		buffer:              newSpansBuffer(spanBufferDefaultMaxSize),
 		sampler:             newAllSampler(),
 		DebugLoggingEnabled: false,
+
+		services:    make(map[string]*Service),
+		serviceChan: make(chan *Service),
```
Comment (LeoCavaille) [inline tracer/tracer.go]: If this is not buffered, the clients might be blocked on `SetServiceInfo` during the flush.

File: tracer/tracer.go [hunk 3]
```diff
@@ -146,16 +186,62 @@ func (t *Tracer) Flush() error {
...
-// worker periodically flushes traces to the transport.
+func (t *Tracer) FlushServices() error {
+	if !t.enabled || !t.servicesModified {
+		return nil
+	}
+
+	if _, err := t.transport.SendServices(t.services); err != nil {
+		return err
+	}
+	log.Printf("sent %d services", len(t.services))
```
Comment (LeoCavaille) [inline tracer/tracer.go]: This should be debug level material so if we can't do it here with the standard logger... maybe drop it?

File: tracer/tracer.go [hunk 4]
```diff
@@ -146,16 +186,62 @@ func (t *Tracer) Flush() error {
...
+				continue mainLoop
+			}
+
+			t.services[service.Name] = service
+			t.servicesModified = true
+
+		case <-t.exit:
+			t.Flush()
+			break mainLoop
 		}
```
Comment (LeoCavaille) [inline tracer/tracer.go]: you can avoid the eye-bothering loop label by doing the opposite test
```go
case service := <-t.serviceChan:
    // skip flushing if the info is still the same
    if s, ok := t.services[service.Name]; !found || !s.Equal(service) {
        t.services[service.Name] = service
        t.servicesModified = true
    }
case <-t.exit:
    t.Flush()
    return
```

File: tracer/transport.go [hunk 5]
```diff
@@ -62,9 +65,11 @@ func newHTTPTransport(hostname, port string) *httpTransport {
 	defaultHeaders["Content-Type"] = contentType
 
 	return &httpTransport{
-		url:       fmt.Sprintf("http://%s:%s/v0.3/traces", hostname, port),
-		legacyURL: fmt.Sprintf("http://%s:%s/v0.2/traces", hostname, port),
-		pool:      pool,
+		traceURL:         fmt.Sprintf("http://%s:%s/v0.3/traces", hostname, port),
+		legacyTraceURL:   fmt.Sprintf("http://%s:%s/v0.2/traces", hostname, port),
+		serviceURL:       fmt.Sprintf("http://%s:%s/v0.3/services", hostname, port),
+		legacyServiceURL: fmt.Sprintf("http://%s:%s/v0.3/services", hostname, port),
```
Comment (LeoCavaille) [inline tracer/transport.go]: v0.2?

File: tracer/tracer.go (hunk 1 above)
Comment (galdor) [inline tracer/tracer.go]: No particular reason, I'll change it.

**Pull request merged**